extras/linked-token/src/LinkedTokenReplica.sol

```diff
@@ -1,28 +1,100 @@
 // SPDX-License-Identifier: MIT OR Apache-2.0
 pragma solidity 0.8.23;
-
-import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
-import {ERC20} from "openzeppelin-contracts/token/ERC20/ERC20.sol";
 import {
-    SafeERC20
-} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol";
+    SafeERC20Upgradeable
+} from "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import {
+    IERC20Upgradeable
+} from "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
 import {LinkedToken} from "./LinkedToken.sol";
 import {SubnetID} from "@ipc/src/structs/Subnet.sol";
+import {
+    Initializable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {
+    ERC20Upgradeable
+} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import {
+    UUPSUpgradeable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 
 /**
  * @title IpcTokenController
  * @notice Contract to handle token transfer from L1, lock them and mint on L2.
  */
-contract LinkedTokenReplica is LinkedToken, ERC20 {
-    using SafeERC20 for IERC20;
+contract LinkedTokenReplica is
+    Initializable,
+    LinkedToken,
+    ERC20Upgradeable,
+    UUPSUpgradeable
+{
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    string _token_name;
+    string _token_symbol;
+    uint8 _token_decimals;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address gateway,
+        address underlyingToken,
+        SubnetID memory linkedSubnet,
+        address linkedContract,
+        string memory token_name,
+        string memory token_symbol,
+        uint8 token_decimals
+    ) public initializer {
+        _token_name = token_name;
+        _token_symbol = token_symbol;
+        _token_decimals = token_decimals;
+
+        __LinkedToken_init(
+            gateway,
+            underlyingToken,
+            linkedSubnet,
+            linkedContract
+        );
+        __UUPSUpgradeable_init();
+        __ERC20_init(_token_name, _token_symbol);
+    }
 
-    constructor(
+    function reinitialize(
```

Comment:
Same comment as above, this enables a double initialization. I think it's incorrect.

Reply:
thanks, fixed https://github.com/consensus-shipyard/ipc/pull/856